[mlir][Linalg] Add transform to convert linalg.copy into memref.copy

pabloantoniom opened the pull request:
Targeted rewrite of a linalg.copy on memrefs to a memref.copy.

This is useful when bufferizing copies to a linalg.copy, later applying some transformations (for instance, tiling), and then rewriting the copy into a memref.copy. If the input linalg.copy has different element type in the source and destination, the transformation is rejected.

pabloantoniom committed:
[mlir][Linalg] Add transform to convert linalg.copy into memref.copy

Targeted rewrite of a linalg.copy on memrefs to a memref.copy.

This is useful when bufferizing copies to a linalg.copy, later applying
some transformations (for instance, tiling), and then rewriting the
copy into a memref.copy. If the input linalg.copy has different element
type in the source and destination, the transformation is rejected.

rolfmorel commented on mlir/test/Dialect/Linalg/transform-op-linalg-copy-to-memref.mlir:
```diff
@@ -0,0 +1,70 @@
+module attributes {transform.with_named_sequence} {
+  transform.named_sequence @__transform_main(%arg1: !transform.any_op {transform.readonly}) {
+    %0 = transform.structured.match ops{["linalg.copy"]} in %arg1 : (!transform.any_op) -> !transform.any_op
+	  // expected-error @below {{linalg.copy on tensors cannot be transformed into memref.copy}}
```
nit: align indentation

pabloantoniom replied:
Thanks for the tip, didn't notice this before...weird stuff from the text editor

rolfmorel commented on mlir/test/Dialect/Linalg/transform-op-linalg-copy-to-memref.mlir:
```diff
@@ -0,0 +1,70 @@
+module attributes {transform.with_named_sequence} {
+  transform.named_sequence @__transform_main(%arg1: !transform.any_op {transform.readonly}) {
+    %0 = transform.structured.match ops{["linalg.copy"]} in %arg1 : (!transform.any_op) -> !transform.any_op
+	  // expected-error @below {{linalg.copy with different source and destination element types is not supported}}
```
nit: align indentation

rolfmorel commented on mlir/test/Dialect/Linalg/transform-op-linalg-copy-to-memref.mlir:
```diff
@@ -0,0 +1,70 @@
+module attributes {transform.with_named_sequence} {
+  transform.named_sequence @__transform_main(%arg1: !transform.any_op {transform.readonly}) {
+    %0 = transform.structured.match ops{["linalg.copy"]} in %arg1 : (!transform.any_op) -> !transform.any_op
+	  // expected-error @below {{linalg.copy with different source and destination element types is not supported}}
```
nit: align indentation

rolfmorel replied:
nit: technically speaking this message is not quite accurate: the issue is with the (lack of) shape(s).

Personally I don't mind too much :shrug:

pabloantoniom replied:
You are right, good catch. I have added a new condition to print an explicit message in this case, and leave the element type error for the case where it's actually because of that.

rolfmorel commented on mlir/include/mlir/Dialect/Linalg/TransformOps/LinalgTransformOps.td:
```diff
@@ -559,6 +559,40 @@ def InterchangeOp : Op<Transform_Dialect, "structured.interchange",
+    Targeted rewrite of a linalg.copy on memrefs to a memref.copy.
+    This is useful when bufferizing copies to a linalg.copy, later applying some
+    transformations, and then rewriting the copy into a memref.copy.
+    If the input has different element type in the source and destination,
```
nit: grammar. Maybe something like "If the `linalg.copy`'s source and destination have distinct element types..."

rolfmorel replied:
Maybe even say that the destination and source types must be the same memref type as the shape and element type must match either way, right?

pabloantoniom replied:
Well, actually no. It should also transform copies where memrefs have strides/offsets (unless this is unsupported by `memref.copy`, please correct me if I'm wrong), so just checking the type would result in non-transformed IR, where it should. Because of this little detail I have also added a new test to demonstrate that this is the intended behaviour and that this works properly.

rolfmorel replied:
Good stuff 👍

rolfmorel reviewed (approved):
Thanks for contributing to the transform dialect, @pabloantoniom! IMO, code quality wise this is very good.

My understanding is that this runs _after_ bufferization as it only applies to linalg-on-memref. As such it shouldn't have (unfortunate) interactions with bufferization. With that in mind, my understanding is that this is a valid (though _very_ gradual) lowering step. It does raise the question for me: is there currently a pass that does this/something similar? If that's so, could you maybe mention it in the summary, i.e. commit message?

I am okay with this to go in with just my nits addressed, though if you could give it another couple days for others to have a look, that would be great. :+1:

pabloantoniom committed:
Fix comments and add new test with strides

pabloantoniom commented on mlir/include/mlir/Dialect/Linalg/TransformOps/LinalgTransformOps.td:
```diff
@@ -559,6 +559,40 @@ def InterchangeOp : Op<Transform_Dialect, "structured.interchange",
+       TransformEachOpTrait, TransformOpInterface]> {
+  let description = [{
+    Targeted rewrite of a linalg.copy on memrefs to a memref.copy.
+    This is useful when bufferizing copies to a linalg.copy, later applying some
```
Done

pabloantoniom commented on mlir/include/mlir/Dialect/Linalg/TransformOps/LinalgTransformOps.td:
```diff
@@ -559,6 +559,40 @@ def InterchangeOp : Op<Transform_Dialect, "structured.interchange",
+    This is useful when bufferizing copies to a linalg.copy, later applying some
+    transformations, and then rewriting the copy into a memref.copy.
+    If the input has different element type in the source and destination,
+    the transformation is rejected.
```
Good point! Done

pabloantoniom commented on mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp:
```diff
@@ -1176,6 +1176,60 @@ LogicalResult transform::InterchangeOp::verify() {
+    transform::ApplyToEachResultList &results,
+    transform::TransformState &state) {
+
+  // Check if the target can be converted
```
Done

pabloantoniom commented on mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp:
```diff
@@ -1176,6 +1176,60 @@ LogicalResult transform::InterchangeOp::verify() {
+  Value output = outputs.front();
+
+  // linalg.copy supports different element types on source/dest whereas
+  // memref.copy does not, so we must check here that the types are the same,
```
Done

pabloantoniom commented:
> Thanks for contributing to the transform dialect, @pabloantoniom! IMO, code quality wise this is very good.
> 
> My understanding is that this runs _after_ bufferization as it only applies to linalg-on-memref. As such it shouldn't have (unfortunate) interactions with bufferization. With that in mind, my understanding is that this is a valid (though _very_ gradual) lowering step. It does raise the question for me: is there currently a pass that does this/something similar? If that's so, could you maybe mention it in the summary, i.e. commit message?
> 
> I am okay with this to go in with just my nits addressed, though if you could give it another couple days for others to have a look, that would be great. 👍

I have changed a couple of small things (mainly rewording) and added a new test. If you like the changes and nobody else has any objection I will merge this shortly. Thanks Rolf! 👍 

Re your question: I'm not aware of something like this.

rolfmorel reviewed (approved):
LGTM
